### PR TITLE
Remove parent and position validation from menu item REST API endpoint

### LINKED ISSUE #25093: REST API: Introduce reorder menu endpoint
**Is your feature request related to a problem? Please describe.**
The Menus endpoint makes it difficult to reorder multiple menu items. This is because the position of each menu item is validated sequentially checking that another menu item doesn't exist with the new position. So if you try to switch the position of two menu items, you'll get an error.

See #21964.

**Describe the solution you'd like**
As suggested by @spacedmonkey, introduce a REST API endpoint specifically for reordering menu items. Where `50` is the menu item id, and `1` is its order.

```http
POST /wp/v2/menus/<menu_id>/ordering
{
      "50": 1
}
```

**Describe alternatives you've considered**
Drop menu item order validation altogether in the endpoint. This isn't validated in `WP_REST_Posts_Controller`. Though based on my reading of `wp_nav_menu`, a duplicate `menu_order` is more of an issue for menus since duplicates will not be printed in the menu.

That being said, AFAICT neither the customizer endpoint nor WP-Admin validate the menu order.

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Description
As noted in https://github.com/WordPress/gutenberg/issues/25093#issuecomment-908709782, we need to remove this validation, so that we can make batch requests work. This PR is simple, remove validation for parent and position. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 

Fixes #25093

### CODE REVIEW

**Comment by talldan** [comment]
Thanks for working on this. Looks like just the PHP Unit tests to update and this should be good to go.

File: phpunit/class-rest-nav-menu-items-controller-test.php
Comment (talldan) [inline phpunit/class-rest-nav-menu-items-controller-test.php]: Should this test be removed? It doesn't seem to trigger validation.

**Review by talldan — APPROVED** [review]
Thanks! I tested alongside @adamziel's PR for making the nav editor use the REST API https://github.com/WordPress/gutenberg/pull/34541, and reordering is working well.

**Comment by talldan** [comment]
Renamed the PR to be a little clearer as the title is what appears in the Gutenberg changelog.

File: phpunit/class-rest-nav-menu-items-controller-test.php
Comment (spacedmonkey) [reply phpunit/class-rest-nav-menu-items-controller-test.php]: It failed.

Comment (adamziel) [reply phpunit/class-rest-nav-menu-items-controller-test.php]: It failed because it resulted in the following sequence of positions: `0, 2, 3, 4, 5`. According to https://core.trac.wordpress.org/ticket/28140, menu position 0 is invalid and breaks things so it would be good to get it fixed. I proposed an adjustment in 4e97c84966 (feel free to roll it back) – what do you think @spacedmonkey ?

**Review by adamziel — APPROVED** [review]

File: phpunit/class-rest-nav-menu-items-controller-test.php
Comment (spacedmonkey) [reply phpunit/class-rest-nav-menu-items-controller-test.php]: I think it might be better change the schema to make the min 1 and default 1. 

https://github.com/WordPress/gutenberg/blob/8e4ebfae09351aa5d21fe072deb16d265bdb130e/lib/class-wp-rest-menu-items-controller.php#L958-L964

Comment (spacedmonkey) [reply phpunit/class-rest-nav-menu-items-controller-test.php]: I would also think about making the position a required field.

**Comment by spacedmonkey** [comment]
Please do not merge until this is resolved - https://github.com/WordPress/gutenberg/pull/34672#discussion_r706260830

File: phpunit/class-rest-nav-menu-items-controller-test.php
Comment (adamziel) [reply phpunit/class-rest-nav-menu-items-controller-test.php]: Nice! Yes, let's do that instead

File: lib/class-wp-rest-menu-items-controller.php
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:528]: Menu id validation is not need here, as this is done upstream.

File: lib/class-wp-rest-menu-items-controller.php
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:925]: This means, that validation, triggers validation.

File: lib/class-wp-rest-menu-items-controller.php
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:714]: Ensure that menus returns a int and not an array. A little unrelated, but good to fix.

File: phpunit/class-rest-nav-menu-items-controller-test.php
Comment (spacedmonkey) [reply phpunit/class-rest-nav-menu-items-controller-test.php]: Fixed.

**Comment by Mamaduka** [comment]
@spacedmonkey, thanks for working on this.

I see the following REST API response error when reordering menu items:

<details>
<summary>Error</summary>

```
{
    "code": "rest_invalid_param",
    "message": "Invalid parameter(s): menu_order",
    "data": {
        "status": 400,
        "params": {
            "menu_order": "menu_order must be greater than or equal to 1"
        },
        "details": {
            "menu_order": {
                "code": "rest_out_of_bounds",
                "message": "menu_order must be greater than or equal to 1",
                "data": null
            }
        }
    }
}
```
</details>

I think we should update `menu_order` to one here as well:
https://github.com/WordPress/gutenberg/blob/c71a933f04d2a234a751d201f19ff4b4d5ee5561/packages/edit-navigation/src/store/actions.js#L68

Current PHP Unit Tests failure also looks like to be related to the recent changes - https://github.com/WordPress/gutenberg/runs/3574827241.

File: lib/class-wp-rest-menu-items-controller.php
Comment (TimothyBJacobs) [reply lib/class-wp-rest-menu-items-controller.php:925]: Are there _any_ cases where this can be `0`? You should generally be able to `PUT` back a `GET` response without any errors.

File: lib/class-wp-rest-menu-items-controller.php
Comment (TimothyBJacobs) [reply lib/class-wp-rest-menu-items-controller.php:528]: Sorry, can you clarify what you mean by "upstream" here?

Comment (spacedmonkey) [reply lib/class-wp-rest-menu-items-controller.php:528]: `menu-id` maps to `menus` when you submit it. Checking it term existings is all handled by [handle_terms](https://github.com/WordPress/wordpress-develop/blob/092ee0fbffaa8723cbddd1134ec811fdcf8601b8/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L1500-L1525). I got the same error. I can return it.

File: lib/class-wp-rest-menu-items-controller.php
Comment (spacedmonkey) [reply lib/class-wp-rest-menu-items-controller.php:925]: Original discussion is here https://github.com/WordPress/gutenberg/pull/34672#discussion_r706133141

**Comment by spacedmonkey** [comment]
@adamziel @TimothyBJacobs @Mamaduka  

There seems to be some confusion here, on if we should not allow 'menu_order' => 0. Seems to lots of backwards and forwards on this one. According to [this](https://core.trac.wordpress.org/ticket/28140), 0 is not a validation position. I am not bother either way. I will defer to y'all on this one.

File: lib/class-wp-rest-menu-items-controller.php
Comment (TimothyBJacobs) [reply lib/class-wp-rest-menu-items-controller.php:528]: Gotcha, that seems fine then to me.

File: lib/class-wp-rest-menu-items-controller.php
Comment (TimothyBJacobs) [reply lib/class-wp-rest-menu-items-controller.php:925]: Ok, let's leave as is for now.

**Review by TimothyBJacobs — APPROVED** [review]
Let's stick with as you have it now @spacedmonkey. I haven't tested the editor for regressions, but if all is working this looks good to merge to me.

**Review by adamziel — APPROVED** [review]

**Pull request merged**
